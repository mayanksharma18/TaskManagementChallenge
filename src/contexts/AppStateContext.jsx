// AppStateContext.js

import React, { createContext, useState, useEffect } from 'react';

const AppStateContext = createContext();

const AppStateProvider = ({ children }) => {
  const [boards, setBoards] = useState([]);

  useEffect(() => {
    // Fetch boards from local storage when the component mounts
    const storedBoards = JSON.parse(localStorage.getItem('boards')) || [];
    setBoards(storedBoards);
  }, []);

  const updateBoards = (newBoards) => {
    setBoards(newBoards);

    // Update local storage
    localStorage.setItem('boards', JSON.stringify(newBoards));
  };

  const addTaskToBoard = (boardName,task) => {
    const updateBoard = [...boards];
    updateBoard.map(boardInLS => {
        if (boardInLS.boardName === boardName) {
            boardInLS.tasks = [...(boardInLS.tasks || []),task]
        }
    })
    updateBoards(updateBoard);
  }

  const addWorkStateToBoard = (boardName,workState) => {
    const updateBoard = [...boards];
    updateBoard.forEach(boardInLS => {
        if (boardInLS.boardName === boardName) {
            boardInLS[workState] = []
        }
    })
    updateBoards(updateBoard);
  }


  const contextValue = {
    boards,
    updateBoards,
    addTaskToBoard,
    addWorkStateToBoard,
  };

  return <AppStateContext.Provider value={contextValue}>{children}</AppStateContext.Provider>;
};

export { AppStateContext, AppStateProvider };
