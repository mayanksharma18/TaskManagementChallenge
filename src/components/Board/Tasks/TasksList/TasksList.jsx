import React, { useContext } from 'react';
import Typography from '@mui/material/Typography';
import { List, ListItem, ListItemText, Stack, Paper, IconButton } from '@mui/material';
import DeleteIcon from '@mui/icons-material/Delete';
import { AppStateContext } from '../../../../contexts/AppStateContext';
import EditTask from '../EditTask/EditTask';

const TasksList = ({ boardName }) => {
  const { boards, updateBoards } = useContext(AppStateContext);

  const tasks = boards.find((board) => board.boardName === boardName)?.tasks || [];

  const handleDeleteTask = (taskId) => {
    const updatedBoards = boards.map((board) => {
      if (board.boardName === boardName) {
        const updatedTasks = board.tasks.filter((task) => task.id !== taskId);
        return { ...board, tasks: updatedTasks };
      }
      return board;
    });

    updateBoards(updatedBoards);
  };

  const handleOnEdit = (updatedTaskDetails) => {
    const updatedBoards = boards.map((board) => {
      if (board.boardName === boardName) {
        const updatedTasks = board.tasks.map((task) => {
          if(task.id === updatedTaskDetails.id){
            return updatedTaskDetails;
          }
          return task
        });
        return { ...board, tasks: updatedTasks };
      }
      return board;
    });

    updateBoards(updatedBoards);

  }

  return (
    <div>
      <Typography variant="h5" gutterBottom>
        Tasks:
      </Typography>
      {tasks.length === 0 ? (
        <Typography>No tasks available.</Typography>
      ) : (
        <List>
          {tasks.map((task) => (
            <Paper
              key={task.id}
              sx={{
                marginBottom: 2,
                padding: 2,
                border: '1px solid #ccc',
                boxShadow: '0 4px 8px rgba(0, 0, 0, 0.1)',
                display: 'flex',
                flexDirection: 'column',
              }}
            >
              <ListItem sx={{ display: 'flex', alignItems: 'center', justifyContent: 'space-between' }}>
                <Stack spacing={1}>
                  <ListItemText primary="Task" secondary={task.taskName} />
                  <ListItemText primary="Description" secondary={task.taskDescription} />
                  <ListItemText primary="Deadline" secondary={task.deadline} />
                  <Stack direction="row" spacing={1}>
                    <IconButton color="primary" size="small" onClick={() => handleDeleteTask(task.id)} data-testid="deleteButton">
                      <DeleteIcon />
                    </IconButton>
                    <EditTask task ={task} handleOnEdit={handleOnEdit}/>
                  </Stack>
                </Stack>
              </ListItem>
            </Paper>
          ))}
        </List>
      )}
    </div>
  );
};

export default TasksList;
