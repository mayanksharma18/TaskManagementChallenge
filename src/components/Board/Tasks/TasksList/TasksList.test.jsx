import React from "react";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import TasksList from "./TasksList";
import { AppStateContext } from "../../../../contexts/AppStateContext";

const mockAppState = {
  boards: [
    {
      boardName: "exampleBoard",
      tasks: [
        {
          id: 1,
          taskName: "Test Task",
          taskDescription: "Test Description",
          deadline: "2023-12-01",
        },
     
      ],
    },
  ],
  updateBoards: jest.fn(),
};

describe("TasksList Component", () => {
  it("render correctly", () => {
    expect(
      render(
        <AppStateContext.Provider value={mockAppState}>
          <TasksList boardName="exampleBoard" />
        </AppStateContext.Provider>
      ).container
    ).toMatchSnapshot();
  });
  
  it("renders TasksList with tasks and handles delete and edit actions", () => {
    render(
      <AppStateContext.Provider value={mockAppState}>
        <TasksList boardName="exampleBoard" />
      </AppStateContext.Provider>
    );
    expect(screen.getByText("Tasks:")).toBeInTheDocument();

    // Expect the task details to be rendered
    expect(screen.getByText("Task")).toBeInTheDocument();
    expect(screen.getByText("Description")).toBeInTheDocument();
    expect(screen.getByText("Deadline")).toBeInTheDocument();

    const deleteButton = screen.getByTestId("deleteButton");
    expect(deleteButton).toBeInTheDocument();

    fireEvent.click(deleteButton);
    waitFor(() => {
      expect(screen.queryByText("Task")).not.toBeInTheDocument();
      expect(screen.queryByText("Description")).not.toBeInTheDocument();
      expect(screen.queryByText("Deadline")).not.toBeInTheDocument();
    });
  });
});
