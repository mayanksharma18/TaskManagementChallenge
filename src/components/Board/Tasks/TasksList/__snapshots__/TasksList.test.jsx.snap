// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksList Component render correctly 1`] = `
<div>
  <div>
    <h5
      class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
    >
      Tasks:
    </h5>
    <ul
      class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-12zae08-MuiPaper-root"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1tbxken-MuiListItem-root"
        >
          <div
            class="MuiStack-root css-jfdv4h-MuiStack-root"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                Task
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
              >
                Test Task
              </p>
            </div>
            <div
              class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                Description
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
              >
                Test Description
              </p>
            </div>
            <div
              class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                Deadline
              </span>
              <p
                class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
              >
                2023-12-01
              </p>
            </div>
            <div
              class="MuiStack-root css-niqf4j-MuiStack-root"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall css-2rshpx-MuiButtonBase-root-MuiIconButton-root"
                data-testid="deleteButton"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <div>
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeSmall css-1hxb263-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="EditIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </li>
      </div>
    </ul>
  </div>
</div>
`;
